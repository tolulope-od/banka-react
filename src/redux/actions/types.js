export const GET_ERRORS = 'GETTING ERRORS';
export const CLEAR_ERRORS = 'CLEARING ERRORS';
export const SET_USER = 'SETTING CURRENT USER';
