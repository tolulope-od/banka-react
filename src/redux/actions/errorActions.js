import { GET_ERRORS, CLEAR_ERRORS } from './types';

export const getErrors = error => ({ type: GET_ERRORS, payload: error });
export const clearErrors = () => ({ type: CLEAR_ERRORS });
