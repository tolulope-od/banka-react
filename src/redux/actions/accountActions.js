import axios from 'axios';
import { ACCOUNT_ACTION_LOADING, ADD_BANK_ACCOUNT } from './types';
import { clearErrors, getErrors } from './errorActions';
import API_URL from '../../utils/API';

const createBankAccount = type => {
  return async dispatch => {
    dispatch({ type: ACCOUNT_ACTION_LOADING, payload: true });
    try {
      dispatch(clearErrors());
      const createdAccount = await axios.post(`${API_URL}/accounts`, { type });
      const { newAccount } = createdAccount.data.data[0];
      return dispatch({ type: ADD_BANK_ACCOUNT, payload: newAccount });
    } catch (err) {
      dispatch({ type: ACCOUNT_ACTION_LOADING, payload: false });
      return dispatch(getErrors(err.response.data));
    }
  };
};

export default createBankAccount;
