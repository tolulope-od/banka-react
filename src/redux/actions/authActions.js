import axios from 'axios';
import jwtDecode from 'jwt-decode';
import setAuthToken from '../../utils/setAuthToken';
import { SET_USER, USER_LOADING } from './types';
import { clearErrors, getErrors } from './errorActions';
import API_URL from '../../utils/API';

export const setUser = decoded => ({ type: SET_USER, payload: decoded });

export const login = userData => {
  return async dispatch => {
    dispatch({ type: USER_LOADING, payload: true });
    try {
      dispatch(clearErrors());
      const returningUser = await axios.post(`${API_URL}/auth/signin`, userData);
      const { token } = returningUser.data.data[0];
      setAuthToken(token);
      const decoded = jwtDecode(token);
      dispatch(setUser(decoded));
      return dispatch({ type: USER_LOADING, payload: false });
    } catch (err) {
      dispatch({ type: USER_LOADING, payload: false });
      return dispatch(getErrors(err.response.data));
    }
  };
};

export const signUp = (userData, history) => {
  return async dispatch => {
    dispatch({ type: USER_LOADING, payload: true });
    try {
      dispatch(clearErrors());
      const newUser = await axios.post(`${API_URL}/auth/signup`, userData);
      const { token } = newUser.data.data[0];
      setAuthToken(token);
      const decoded = jwtDecode(token);
      dispatch(setUser(decoded));
      dispatch({ type: USER_LOADING, payload: false });
      return history.push('/createAccount');
    } catch (err) {
      dispatch({ type: USER_LOADING, payload: false });
      return dispatch(getErrors(err.response.data));
    }
  };
};
