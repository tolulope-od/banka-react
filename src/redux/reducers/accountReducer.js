import { ACCOUNT_ACTION_LOADING, ADD_BANK_ACCOUNT } from '../actions/types';

export const initialState = {
  account: {},
  accounts: [],
  accountLoading: false
};

export default (state = initialState, action) => {
  switch (action.type) {
    case ACCOUNT_ACTION_LOADING:
      return {
        ...state,
        accountLoading: action.payload
      };
    case ADD_BANK_ACCOUNT:
      return {
        ...state,
        account: action.payload,
        accounts: [action.payload, ...state.accounts],
        accountLoading: false
      };
    default:
      return state;
  }
};
