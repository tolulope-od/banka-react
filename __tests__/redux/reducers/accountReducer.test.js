import accountReducer, { initialState } from '../../../src/redux/reducers/accountReducer';
import { ACCOUNT_ACTION_LOADING, ADD_BANK_ACCOUNT } from '../../../src/redux/actions/types';

describe('Account Reducer Tests', () => {
  it('should return the initial state', () => {
    expect(accountReducer(undefined, {})).toMatchSnapshot();
  });

  it('Should set the loading state when account information is being requested', () => {
    const payload = true;
    expect(accountReducer(initialState, { type: ACCOUNT_ACTION_LOADING, payload })).toEqual({
      account: {},
      accountLoading: true,
      accounts: []
    });
  });

  it('should add a newly created account to the state', () => {
    const payload = {
      id: 35,
      accountnumber: '6407771369',
      owner: 5,
      ownername: 'Joy Machido',
      type: 'savings',
      status: 'active',
      balance: '0.0',
      owneremail: 'joy.machido@gmail.com',
      createdon: '2019-08-20T13:22:41.935Z'
    };
    expect(accountReducer(initialState, { type: ADD_BANK_ACCOUNT, payload })).toEqual({
      account: payload,
      accountLoading: false,
      accounts: [payload]
    });
  });
});
