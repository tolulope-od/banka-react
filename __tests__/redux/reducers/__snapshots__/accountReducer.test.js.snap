// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account Reducer Tests should return the initial state 1`] = `
Object {
  "account": Object {},
  "accountLoading": false,
  "accounts": Array [],
}
`;
